### ISSUE TITLE
Azure Workload Identity - Expired token

### ISSUE THREAD

**Issue by hjarraya** [Azure/azure-sdk-for-rust#1739]
We are running Azure Identity v0.20.0 and getting this error for a longer-running process in AKS. We are using Workload Identity with our deployments. 

```
 azure_identity::federated_credentials_flow: rsp_body == "{
 "error":"invalid_client",
 "error_description":"AADSTS700024: Client assertion is not within its valid time range.  
   Current time: 2024-08-05T10:07:04.2542067Z, assertion 
   valid from 2024-08-02T18:52:57.0000000Z, expiry time of assertion 2024-08-02T19:52:57.0000000Z.  Review the documentation at https://learn.microsoft.com/entra/identity-platform/certificate-credentials . Trace ID: XXXX-XXXX-XXXX-XXXX-XXXX Correlation ID: XXXX-XXXX-XXXX-XXXX-XXXX Timestamp: 2024-08-05 10:07:04Z",
   "error_codes":[700024],
   "timestamp":"2024-08-05 10:07:04Z", "trace_id":"XXXX-XXXX-XXXX-XXXX-XXXX",
   "correlation_id": "XXXX-XXXX-XXXX-XXXX-XXXX","error_uri": "https://login.microsoftonline.com/error?code=700024"
   }"
``` 

The workload identity token is not being refreshed as it is supposed to be.

**Comment by erewok** [comment]
I have the same problem. This is occurring on `0.20.0` and did not occur with the previous version of the library: `0.19.0`.

**Comment by eplightning** [comment]
Looks like the federated token (e.g. Kubernetes SA projected token) is only loaded once since `0.20.0`. 

`0.19.0` `EnvironmentCredential` loaded it every time so it just worked back then:
```
            let token = std::fs::read_to_string(file.clone())
                .with_context(ErrorKind::Credential, || {
                    format!("failed to read federated token from file {}", file.as_str())
                })?;
            let mut credential: WorkloadIdentityCredential = WorkloadIdentityCredential::new(
                self.http_client.clone(),
                tenant_id,
                client_id,
                token,
            );
            credential.set_options(options);

            return credential.get_token(scopes).await;
```

**Comment by hjarraya** [comment]
Indeed, it is loaded every time, so I cached it locally and requested it when it was about to expire.

**Comment by inganault** [comment]
This was caused by WorkloadIdentityCredential caching AZURE_FEDERATED_TOKEN_FILE file **content** introduced in v0.20.
https://github.com/Azure/azure-sdk-for-rust/blob/6cf4cc9f08a814f8d88d75e988c6f6f2559fb564/sdk/identity/azure_identity/src/credentials/workload_identity_credentials.rs#L103

The fix is to store only the file path and reading them every time like [azure python sdk](https://github.com/Azure/azure-sdk-for-python/blob/7fd7af86bc43ee83b2e22cd8b054dba8f458de7a/sdk/identity/azure-identity/azure/identity/_credentials/workload_identity.py#L23).

**Comment by arpad-m** [comment]
I guess the PR that caused the regression was #1532 , cc @heaths .

https://github.com/Azure/azure-sdk-for-rust/pull/1532/files#diff-2ad4468fd1fd66aac39f9410601f0b67dd7d796bf9a36db12c5b6a87c0b99d35R93

**Comment by heaths** [comment]
@arpad-m are you able to test #1997 against your fork? I'm working on our live / recorded / mock testing now and don't have a good environment in which to test this, but sounds like you do. We'll be adding a lot of integration tests once my current work is finished.

**Comment by arpad-m** [comment]
@heaths thanks for the quick reply! I would happily try out your patch in our staging environment where this issue occurs. I've filed https://github.com/neondatabase/neon/pull/10415 to start the process, and asked my colleagues about it. I'll keep in touch. I'll try to tell you about the test's outcome in the coming days.

**Comment by arpad-m** [comment]
@heaths we have tried your PR again after improvements with our staging set up, and now we can say that your PR has fixed the problem for us.
